src/picongpu/include/fields/FieldJ.kernel

```diff
@@ -46,80 +48,233 @@ using namespace PMacc;
+ */
+template<
+    uint32_t T_numWorkers,
+    typename T_BlockDescription
+>
 struct KernelComputeCurrent
 {
-    template<class JBox, class ParBox, class Mapping, class FrameSolver>
-    DINLINE void operator()(JBox fieldJ,
-                                         ParBox boxPar, FrameSolver frameSolver, Mapping mapper) const
+    /** scatter particle current of particles located in a supercell
+     *
+     * The current for the supercell including the guards is cached in shared memory
+     * and scattered at the end of the functor to the global memory.
+     *
+     * @tparam JBox particle current box type
+     * @tparam ParBox particle box type
+     * @tparam Mapping mapper functor type
+     * @tparam FrameSolver frame solver functor type
+     *
+     * @param fieldJ field with particle current (the interface of a PMacc::DataBox must be fulfilled)
+     * @param boxPar particle memory (the interface of a PMacc::ParticlesBox must be fulfilled)
+     * @param frameSolver functor to calculate the current for a frame
+     * @param mapper functor to map a block to a supercell
+     */
+    template<
+        typename JBox,
+        typename ParBox,
+        typename Mapping,
+        typename FrameSolver
+    >
+    DINLINE void operator()(
+        JBox fieldJ,
+        ParBox boxPar,
+        FrameSolver frameSolver,
+        Mapping mapper
+    ) const
     {
-        typedef typename ParBox::FrameType FrameType;
-        typedef typename ParBox::FramePtr FramePtr;
-        typedef typename Mapping::SuperCellSize SuperCellSize;
-
-        const uint32_t cellsPerSuperCell = PMacc::math::CT::volume<SuperCellSize>::type::value;
-
-        const DataSpace<simDim> block(mapper.getSuperCellIndex(DataSpace<simDim > (blockIdx)));
-        const DataSpace<simDim > threadIndex(threadIdx);
-
-        /* thread id, can be greater than cellsPerSuperCell*/
-        const int linearThreadIdx = DataSpaceOperations<simDim>::template map<SuperCellSize > (threadIndex);
-
-        const uint32_t virtualBlockId = linearThreadIdx / cellsPerSuperCell;
-        /* move linearThreadIdx for all threads to [0;cellsPerSuperCell) */
-        const int virtualLinearId = linearThreadIdx - (virtualBlockId * cellsPerSuperCell);
-
-
-        FramePtr frame;
-        lcellId_t particlesInSuperCell = 0;
-
-        frame = boxPar.getLastFrame(block);
-        if (frame.isValid() && virtualBlockId == 0)
-            particlesInSuperCell = boxPar.getSuperCell(block).getSizeLastFrame();
+        using namespace mappings::threads;
+
+        using FrameType = typename ParBox::FrameType;
+        using FramePtr = typename ParBox::FramePtr;
+        using SuperCellSize = typename Mapping::SuperCellSize;
+
+        /** @todo numParticlesPerFrame should be max number of particles within a frame
+         * and not a magic number derived from SuperCellSize
+         */
+        constexpr uint32_t numParticlesPerFrame = PMacc::math::CT::volume< SuperCellSize >::type::value;
```

Comment:
@ax3l  This is not possible in this pull request because this must be given from PIConGPU to PMacc, this means a lot of template signature changes and all algorithms needs to take care of the two domains ( superCells and particles per frame)